src/storage/projected_row.cpp

```diff
@@ -64,4 +60,44 @@ ProjectedRow *ProjectedRowInitializer::InitializeRow(void *const head) const {
   result->Bitmap().Clear(result->num_cols_);
   return result;
 }
+
+ProjectedRowInitializer ProjectedRowInitializer::CreateProjectedRowInitializer(const BlockLayout &layout,
+                                                                               std::vector<col_id_t> col_ids) {
+  TERRIER_ASSERT(col_ids.size() < layout.NumColumns(),
+                 "ProjectedRow should have fewer columns than the table (can't read version vector)");
+  // Sort the projection list for optimal space utilization and delta application performance
+  // If the col ids are valid ones laid out by BlockLayout, ascending order of id guarantees
+  // descending order in attribute size.
+  std::sort(col_ids.begin(), col_ids.end(), std::less<>());
+  std::vector<uint8_t> attr_sizes;
+  attr_sizes.reserve(col_ids.size());
+  for (auto const &col_id : col_ids) {
+    attr_sizes.emplace_back(layout.AttrSize(col_id));
+  }
+  return ProjectedRowInitializer(attr_sizes, std::move(col_ids));
+}
+
+template <typename AttrType>
+ProjectedRowInitializer ProjectedRowInitializer::CreateProjectedRowInitializerForIndexes(
+    std::vector<AttrType> attr_sizes, const std::vector<uint16_t> &pr_offsets) {
+  // uint8 requires masking off the MSB varlen flag, uint16 are actual sizes
```

Comment:
Why is the MSB flipped for uint8?

Reply:
Refactored into an additional method GetRealAttrSizes() in IndexMetadata, which we call before invoking CreateProjectedRowInitializerForIndexes(). The MSB was flipped because the uint8_t's use the MSB to indicate whether a column is VARLEN or not. Now, the initializer always expects the real attribute sizes.